db_index: run `initial_scan` once

QuerthDP opened the pull request:
When received a message during initial scanning, the `initial_scan` future was built every time again.
Due to the bug, the completed_scan_length counter was incremented to many times, leading the counter to overflow, which lead into the infinite loop described in the Jira ticket.

Now it's built outside the select once and is still listened to concurrently with listening for messages.

Fixes: VECTOR-192

QuerthDP committed:
db_index: run `initial_scan` once

When recieved a message during initial scanning, the `initial_scan`
future was built every time again.
Due to the bug, the completed_scan_length counter was incremented
to many times, leading the counter to overflow, which lead into the
infinite loop described in the Jira ticket.

Now it's built outside the select once and is still listened to
concurrently with listening for messages.

Fixes: VECTOR-192

knowack1 commented:
> I wanted to add tests to this PR, but I believe the [VECTOR-114](https://scylladb.atlassian.net/browse/VECTOR-114) is required to accomplish this. Maybe I'm wrong, I will look into it further.

Nice investigation. The problem is really tricky.
Indeed reliable test requires control over range scanning - this driver call:
```
.session
            .execute_iter(self.st_range_scan.clone(), (begin.value(), end.value()))
```

QuerthDP committed:
validator/common: add `wait_for` test function

Add `wait_for` test function to enable waiting for a task
to finish under a certain condition. Otherwise timeout.

QuerthDP committed:
httpclient: refactor the `/count` endpoint calling function

Adjust a function to send GET request to `/count` endpoint
using the keyspace and index name as an arguments.

It now does not need the `IndexMetadata` structure and can
be used more widely if only the keyspace and index name are
accessible.

QuerthDP committed:
validator/test: add full scan test

Test if sending ANN request while full scan is being conducted
does not break the full scan.

swasik commented:
> And as the Validator test failed it seems that the test does not work on `scylladb/scylladb-releng:2025.4.0-dev-0.20250718.d36c9b3119b7-x86_64` on which the CI is run. However it does work on the newest build (`scylladb/scylla-nightly` for example). The difference comes from that the error message is more descriptive now since [scylladb/scylladb#24444](https://github.com/scylladb/scylladb/pull/24444) was merged.

We had plenty of changes merged recently - I think it would be worth to bump the version used by validator (in a separate PR of course).

QuerthDP commented:
Changelog:

- rebased on master
- added descriptive commit messages

ewienik commented:
> Validator tests should fail due to [VECTOR-197](https://scylladb.atlassian.net/browse/VECTOR-197).

The problem could be with DNS TTL - currently it is set to 60s: https://github.com/scylladb/vector-store/blob/master/crates/validator/src/dns.rs#L119. I think it could be solved with this PR. Check with TTL set to 1, and add a async sleep for TTL in upsert.

QuerthDP commented:
Changelog:

- refactored the httpclient `count` function not to use `IndexMetadata` object
- made the `wait_for` timeout adjustable as a function parameter

QuerthDP commented:
Changelog:

- rebased on master
